measures/src/Data/Measure/ByteSize.hs

```diff
@@ -0,0 +1,328 @@
+-- They intentionally have no operators! All calculations should be done in
+-- 'ByteSize'.
+newtype ByteSize8 = ByteSize8 { unByteSize8 :: Word8 }
+    -- ^ See the type's Haddock.
+  deriving stock   (Generic)
+  deriving stock   (Eq, Ord)
+  deriving newtype (Bounded)
+  deriving newtype (NFData)
+  deriving newtype (FromCBOR, ToCBOR)
+  deriving         (Read, Show)
+           via         Quiet ByteSize8
+  deriving         (NoThunks)
+           via         OnlyCheckWhnfNamed "ByteSize8" ByteSize8
+
+-- | See the documentation on 'ByteSize8'.
+newtype ByteSize16 = ByteSize16 { unByteSize16 :: Word16 }
+    -- ^ See the documentation on 'ByteSize8'.
+  deriving stock   (Generic)
+  deriving stock   (Eq, Ord)
+  deriving newtype (Bounded)
+  deriving newtype (NFData)
+  deriving newtype (FromCBOR, ToCBOR)
+  deriving         (Read, Show)
+           via         Quiet ByteSize16
+  deriving         (NoThunks)
+           via         OnlyCheckWhnfNamed "ByteSize16" ByteSize16
+
+-- | See the documentation on 'ByteSize8'
+newtype ByteSize32 = ByteSize32 { unByteSize32 :: Word32 }
+    -- ^ See the documentation on 'ByteSize8'.
+  deriving stock   (Generic)
+  deriving stock   (Eq, Ord)
+  deriving newtype (Bounded)
+  deriving newtype (NFData)
+  deriving newtype (FromCBOR, ToCBOR)
+  deriving         (Read, Show)
+           via         Quiet ByteSize32
+  deriving         (NoThunks)
+           via         OnlyCheckWhnfNamed "ByteSize32" ByteSize32
+
+-- | See the documentation on 'ByteSize8'
+newtype ByteSize64 = ByteSize64 { unByteSize64 :: Word64 }
+    -- ^ See the documentation on 'ByteSize8'.
+  deriving stock   (Generic)
+  deriving stock   (Eq, Ord)
+  deriving newtype (Bounded)
+  deriving newtype (NFData)
+  deriving newtype (FromCBOR, ToCBOR)
+  deriving         (Read, Show)
+           via         Quiet ByteSize64
+  deriving         (NoThunks)
+           via         OnlyCheckWhnfNamed "ByteSize64" ByteSize64
+
+--------------------------------------------------------------------------------
+-- Conversions
+--------------------------------------------------------------------------------
+
+-- | Types where every value can be decidably classified as one of the following.
+--
+--   - An integer within the INVARIANT interval of 'ByteSize'.
+--   - An integer greater than the INVARIANT interval of 'ByteSize'.
+--   - Neither of those. For example, negative numbers, fractions, orange, etc.
+--
+-- Law: 'partialFrom' returns 'SNothing' only in the third case above.
+--
+-- Law: 'partialFrom' returns 'SJust' an overflow if and only if the second
+-- case above.
+class ByteSizePartialFrom a where
+    -- | See 'ByteSizePartialFrom'.
+    partialFrom :: a -> StrictMaybe ByteSize
+
+partialFromDefault :: ByteSizeFrom a => a -> StrictMaybe ByteSize
+partialFromDefault = SJust . from
+
+-- | Law: @'partialFrom' = 'SJust' . 'from'@
+class ByteSizePartialFrom a => ByteSizeFrom a where
+    -- | See 'ByteSizeFrom'.
+    from :: a -> ByteSize
```

Comment:
Not only `to` and `from` are very confusing and ambiguous names, but they are also highly likely to be used in user's codebase. Without qualification these will lead to clashes quite often (even `lens`/`microlens` has `to`)

I'd suggest better names that depict what is the actual conversion here. So in order to avoid ambiguity of what is being converted to what, I'd suggest something like this:

```haskell
-- | A type that can be safely converted to `ByteSize`
class ToByteSize a where
    toByteSize :: a -> ByteSize

-- | A type that can be safely constructed from `ByteSize`
class FromByteSize a where
    fromByteSize :: a -> ByteSize
```

Reply:
> Without qualification ...

So far, I'm intending for this module to be imported qualified.

Reply:
Do we need to export `SJust . from` at all?

Reply:
> So far, I'm intending for this module to be imported qualified.

I appreciate your intention, but that should not be forced on the user. Moreover, qualification does not solve the ambiguity of naming `from` and `to`.

So, to put it bluntly, I will not let functions with names like `from` and `to` to be merged.